src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,211 @@
+        {
+            // check that the user is not typing in a string literal or comment
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
+        }
+
+        public async Task<BraceCompletionContext?> GetCompletedBraceContextAsync(Document document, int caretLocation, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var leftToken = root.FindTokenOnLeftOfPosition(caretLocation);
+            var rightToken = root.FindTokenOnRightOfPosition(caretLocation);
+
+            if (IsValidOpeningBraceToken(leftToken) && IsValidClosingBraceToken(rightToken))
+            {
+                return new BraceCompletionContext(document, leftToken.GetLocation().SourceSpan.Start, rightToken.GetLocation().SourceSpan.End, caretLocation);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Checks if the already inserted token is a valid opening token at the position in the document.
+        /// By default checks that the opening token is a valid token at the position and not in skipped token trivia.
+        /// </summary>
+        protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            // The open token is typed in skipped token trivia, we should not attempt to complete it.
+            if (ParentIsSkippedTokensTrivia(syntaxFactsService, token))
+            {
+                return SpecializedTasks.False;
+            }
+
+            return Task.FromResult(token.SpanStart == position && IsValidOpeningBraceToken(token));
+        }
+
+        /// <summary>
+        /// Returns true when the current position is inside user code (e.g. not strings) and the closing token
+        /// matches the expected closing token for this brace completion service.
+        /// Helper method used by <see cref="AllowOverTypeAsync(BraceCompletionContext, CancellationToken)"/> implementations.
+        /// </summary>
+        protected async Task<bool> AllowOverTypeInUserCodeWithValidClosingTokenAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            var tree = await context.Document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = context.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return !syntaxFactsService.IsInNonUserCode(tree, context.CaretLocation, cancellationToken)
```

Comment:
would it make sense at all to push the:

1. am i on a brace?
2. am i in non-user code?

all the way up to the top, so that this check is only in one plkace?

Reply:
These checks have to happen here because there are two entry points (which are triggered by two different locations in the users code for two different use cases).  They're basically at the 'top' already.
1.  Determining if this brace completion service can be used to complete a brace.  This also checks the opening _character_ (not inserted).
2.  Determining if overtyping is allowed at a location (and the non-user code check only applies to certain kinds of braces).  This also checks the closing _token_ kind (inserted)